File: 0/1/7/017.hs
-- Copyright (C) 2013 Jorge Aparicio

main :: IO()
main
  = print
  . sum
  . map wordify
  $ [1..1000]

wordify :: Int -> Int
wordify 1 = length "one"
wordify 2 = length "two"
wordify 3 = length "three"
wordify 4 = length "four"
wordify 5 = length "five"
wordify 6 = length "six"
wordify 7 = length "seven"
wordify 8 = length "eight"
wordify 9 = length "nine"
wordify 10 = length "ten"
wordify 11 = length "eleven"
wordify 12 = length "twelve"
wordify 13 = length "thirteen"
wordify 15 = length "fifteen"
wordify 18 = length "eighteen"
wordify 20 = length "twenty"
wordify 30 = length "thirty"
wordify 40 = length "forty"
wordify 50 = length "fifty"
wordify 80 = length "eighty"
wordify 1000 = wordify 1 + length "thousand"
wordify n
  | n < 20    = wordify u + length "teen"
  | n < 100   = if u == 0
                   then wordify (t `div` 10) + length "ty"
                   else wordify t + wordify u
  | otherwise = if tau == 0
                   then wordify (h `div` 100) + length "hundred"
                   else wordify h + length "and" + wordify tau
    where tau = n `mod` 100
          u = n `mod` 10
          t  = tau - u
          h = n - tau

File: 0/1/7/017.rs
// Copyright (C) 2014 Jorge Aparicio

use std::iter::AdditiveIterator;

fn main() {
  println!("{}", range(1u, 1001).map(|n| wordify(n)).sum());
}

fn wordify(n: uint) -> uint {
  if n < 10 {
    match n {
      1 => "one".len(),
      2 => "two".len(),
      3 => "three".len(),
      4 => "four".len(),
      5 => "five".len(),
      6 => "six".len(),
      7 => "seven".len(),
      8 => "eight".len(),
      _ => "nine".len()
    }
  } else if n < 20 {
    match n {
      10 => "ten".len(),
      11 => "eleven".len(),
      12 => "twelve".len(),
      13 => "thirteen".len(),
      15 => "fifteen".len(),
      18 => "eighteen".len(),
      _ => wordify(n % 10) + "teen".len()
    }
  } else if n < 100 {
    let u = n % 10;
    if u == 0 {
      match n {
        20 => "twenty".len(),
        30 => "thirty".len(),
        40 => "forty".len(),
        50 => "fifty".len(),
        80 => "eighty".len(),
        _ => wordify(n / 10) + "ty".len()
      }
    } else {
      let t = n - u;
      wordify(t) + wordify(u)
    }
  } else if n < 1000 {
    let tau = n % 100;
    if tau == 0 {
      wordify(n / 100) + "hundred".len()
    } else {
      let h = n - tau;
      wordify(h) + "and".len() + wordify(tau)
    }
  } else {
    wordify(n / 1000) + "thousand".len()
  }
}

